### PR TITLE
v3 Release: NFT + code cleaning + doc update

### PR DESCRIPTION
**Pull request by kalwalt**
**⚠️ All PRs have to be done versus 'dev' branch, so be aware of that, or we'll close your issue ⚠️**
From https://github.com/jeromeetienne/AR.js/pull/674#issue-352547279 #674 
**What kind of change does this PR introduce?**
This PR want to test the new jsartoolkit5 from https://github.com/kalwalt/jsartoolkit5/pull/1 with the new changes especially the NFT part.

### **🖼 This will eventually make AR.js capable of detect and track 2D images.**

https://twitter.com/nicolocarp/status/1204841439160684549

Could be raise some bugs and incompatibilties between the code. It is more a test...
**Can it be referenced to an Issue? If so what is the issue # ?**

From jsartoolkit5 repo:  PR  https://github.com/kalwalt/jsartoolkit5/pull/1
From AR.js issues:

* https://github.com/jeromeetienne/AR.js/issues/544
* https://github.com/jeromeetienne/AR.js/issues/149
* https://github.com/jeromeetienne/AR.js/issues/503
* https://github.com/jeromeetienne/AR.js/issues/420
* https://github.com/jeromeetienne/AR.js/issues/227
* https://github.com/jeromeetienne/AR.js/issues/190
* https://github.com/jeromeetienne/AR.js/issues/185

also, a repo cleaning will help for: https://github.com/jeromeetienne/AR.js/issues/433

**How can we test it?**
Just test the examples and tell us what happens...

**Summary**

**Does this PR introduce a breaking change?**
Nope, it will remove old code, but unnecessary. It will add NFT feature.

**Please TEST your PR before proposing it. Specify here what device you have used for tests, version of OS and version of Browser**

**Other information**
I uploaded the new jsartookit5 libs and build the AR.js libs with makefile. 

**Progress**
- [x] Clean repo before starting the porting
- [x] Import new jsartoolkit5 version, coming with NFT
- [x] Adapt the code (threejs side) to support NFT (emulate what's already done for fiducial markers)
- [x] Fix the rendering of content on AR.js side, for NFT (onMarkerFound function)
- [x] Make it work on mobile
- [ ] Enhance resolution in both mobile and desktop
- [ ] Check if animations work with NFT. If not, or rendering rate is too high, use interpolation
- [ ] Fix broken pattern marker (content appears shifted after NFT modifies, or maybe even before during the cleaning of the repo)
- [x] Remove all rendering code on the Worker
- [x] Adapt the code (aframe side) to support NFT (emulate what's already done for fiducial markers)
- [x] Create at least one example for threejs + NFT
- [x] Create at least one example for aframe + NFT
- [ ] Cable the change of position of the 3D object (x = 100, y = 100) inside the core, so we don't have to replicate it in every example (both aframe/three.js)
- [ ] Change documentation (principal Readme, also with NFT Marker generator + Readmes inside threejs + aframe folders)
- [ ] release version 3 of AR.js

### CODE REVIEW

**Comment by kalwalt** [comment]
Now the **threejs/examples/nft,nft** display the sphere onto the pinball marker (image): that is **GREAT**! :smile: :raised_hands: :tada: 
few issues to note:

- in all examples with Pattern Markers cube appears shifted
- default-thinner-border.html example i get this error:
```
aframe-ar.js:1424 Uncaught TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.
    at onSourceReady (aframe-ar.js:1424)
    at ARjs.Source.THREEx.ArToolkitSource.ARjs.Source._initSourceImage (aframe-ar.js:1452)
    at ARjs.Source.THREEx.ArToolkitSource.ARjs.Source.init (aframe-ar.js:1404)
    at new ARjs.Session (aframe-ar.js:2490)
    at HTMLElement.<anonymous> (aframe-ar.js:4500)
    at HTMLElement.emit (a-node.js:263)
    at HTMLElement.<anonymous> (a-scene.js:675)
    at HTMLElement.emit (a-node.js:263)
    at a-node.js:128
```
- aframe example with nft not load the marker ( i need to look into that...)

**Comment by kalwalt** [comment]
Tested also on a Smartphone device: it is difficult to detect the pinball image, it detected on landscape mode with difficulty. As we talked @nicolocarpignoli we should fix the aspect ratio.

File: three.js/Makefile [hunk 1]
```diff
@@ -17,7 +17,12 @@ minify-main: build-main
 
 build: build-main
 
-minify: minify-main
+minify-lean: build-lean
```
Comment (kalwalt) [inline three.js/Makefile]: i think we should delete the lean part again, if i'm not wrong we deleted because not necessary anymore.

Comment (nicolocarpignoli) [reply three.js/Makefile]: ok sorry I re-added for mistake. PLease delete

**Comment by nicolocarpignoli** [comment]
> Now the **threejs/examples/nft,nft** display the sphere onto the pinball marker (image): that is **GREAT**! 😄 🙌 🎉
> few issues to note:
> 
> * in all examples with Pattern Markers cube appears shifted
> * **default-thinner-border.html** example (aframe and three.js) i get this error:
> 
> ```
> aframe-ar.js:1424 Uncaught TypeError: Failed to execute 'appendChild' on 'Node': parameter 1 is not of type 'Node'.
>     at onSourceReady (aframe-ar.js:1424)
>     at ARjs.Source.THREEx.ArToolkitSource.ARjs.Source._initSourceImage (aframe-ar.js:1452)
>     at ARjs.Source.THREEx.ArToolkitSource.ARjs.Source.init (aframe-ar.js:1404)
>     at new ARjs.Session (aframe-ar.js:2490)
>     at HTMLElement.<anonymous> (aframe-ar.js:4500)
>     at HTMLElement.emit (a-node.js:263)
>     at HTMLElement.<anonymous> (a-scene.js:675)
>     at HTMLElement.emit (a-node.js:263)
>     at a-node.js:128
> ```
> 
> * aframe example with nft not load the marker ( i need to look into that...)

ok don't bother nft on aframe for now.

> * in all examples with Pattern Markers cube appears shifted

this is something to fix. I added on checklist

**Comment by nicolocarpignoli** [comment]
I updated the Checklist on Readme. We should follow it and try to fix point by point

File: three.js/Makefile (hunk 1 above)
Comment (kalwalt) [reply three.js/Makefile]: ok i will delete this.

File: three.js/src/threex/threex-armarkercontrols.js [hunk 2]
```diff
@@ -266,10 +267,10 @@ ARjs.MarkerControls.prototype._initArtoolkit = function () {
         var sw, sh;
         var pscale, sscale;
         var w, h;
-        var pw, ph;
         var ox, oy;
-        vw = input_width; // need to setup input video width and height
-        vh = input_height;
+
+        vw = input_width = 640; // need to setup input video width and height
```
Comment (kalwalt) [inline three.js/src/threex/threex-armarkercontrols.js]: This need to be fixed. I'm trying to add an EventListener to retrieve the video width and height as we did here https://github.com/kalwalt/AR.js/blob/2dd0dc8d2c98122cd7998103921f6ff95aa36fe9/three.js/src/threex/threex-armarkercontrols.js#L276-L283 but i get issues with video data.

File: three.js/src/threex/threex-armarkercontrols.js [hunk 3]
```diff
@@ -1,7 +1,7 @@
 var ARjs = ARjs || {}
 var THREEx = THREEx || {}
 
-ARjs.MarkerControls = THREEx.ArMarkerControls = function (context, object3d, parameters) {
+ARjs.MarkerControls = THREEx.ArMarkerControls = function (context, object3d, camera, parameters) {
```
Comment (kalwalt) [inline three.js/src/threex/threex-armarkercontrols.js]: We can find another solution @nicolocarpignoli ?

File: three.js/src/threex/threex-armarkercontrols.js (hunk 2 above)
Comment (nicolocarpignoli) [reply three.js/src/threex/threex-armarkercontrols.js]: what you mean?

I guess the video input can be the video element? So the solution can be to take its sizes

File: three.js/src/threex/threex-armarkercontrols.js (hunk 3 above)
Comment (nicolocarpignoli) [reply three.js/src/threex/threex-armarkercontrols.js]: in which sense?

File: three.js/src/threex/threex-armarkercontrols.js (hunk 2 above)
Comment (kalwalt) [reply three.js/src/threex/threex-armarkercontrols.js]: you are right, much simpler...

File: three.js/src/threex/threex-armarkercontrols.js (hunk 3 above)
Comment (kalwalt) [reply three.js/src/threex/threex-armarkercontrols.js]: in the sense  i need the camera object as function args , but i think i have no choice.

**Comment by kalwalt** [comment]
a little step: with the latest https://github.com/jeromeetienne/AR.js/pull/705/commits/d93ad9cd0a5dab7e1b632d4f61961ba2cb7f5d91 the sphere appears also on Mobile but only in landscape mode. We need to fix this, and also the sphere is flashing at the beginning.

**Comment by nicolocarpignoli** [comment]
@Carnaux are you able to modify the Readme, adding a section for NFT and also for a part for the NFT Marker Generator?
It will help us a lot, it's one of the point of the checklist in the first post :)

**Comment by Carnaux** [comment]
@nicolocarpignoli Yes!

**Comment by Marcus5234** [comment]
Any ETA on this? When do you guys think image tracking will be available in arjs?

**Comment by kalwalt** [comment]
> Any ETA on this? When do you guys think image tracking will be available in arjs?

Very difficult to say, but me and @nicolocarpignoli  are very motivated that this happens. actually we have an issue testing on smartphone device: while on landscape mode it detect and track the NFT marker, instead in portrait mode this not  happens. Probably an iusse with ratio or related thing, but hard to sort out for the Ar.js complexity. Anyway we hope to solve this soon !! :smile: :tada:

**Comment by nicolocarpignoli** [comment]
If anyone is will to help is very welcome, to speed up the process

**Comment by nicolocarpignoli** [comment]
It now works on aframe side and three.js side (after mergin v3-working branch).
Updated the checklist.
Work in progress 💪🏻

**Comment by kalwalt** [comment]
That's a great Progress indeed, let's improve it now!

**Comment by nicolocarpignoli** [comment]
Can you please re-open this PR on the new dev branch? thank you

**Comment by nicolocarpignoli** [comment]
sorry had to clean the dev branch

**Pull request closed without merging**